endpoint: Place IngressIPs to endpoints map

jrajahalme opened the pull request:
Cilium agent manages Ingress endpoint for the purpose of policy generation for Cilium Ingress. The Ingress endpoint has IPs from the node's CIDR range like any other endpoint, but it does not have any representation at the bpf datapath, as it is only used for policy enforcement and for IP addressing at the Envoy proxy. Due to the lacking bpf datapath representation IPv6 ND does not currently work for the Ingress IPs, which causes IPv6 communication with backends not working.

Fix this by introducing minimal bpf datapath representation for the Ingress endpoint via Ingress endpoint entries in the lxcmap, which is used by the IPV6 ND implementation in the Cilium bpf datapath to enable ND advertisements to be sent when ND request for the Ingress IPv6 has been received.

A new endpoint flag ENDPOINT_F_ATHOSTNS is added to inform the datapath to deliver endpoint's traffic to the host stack, but for which ARP and IPv6 ND are still served by the bpf datapath due to the endpoint's IP being from the node's IP allocation CIDR range.

Fixes: #32980

```release-note
Ingress endpoint is now included in the lxcmap so that ARP and ND6 work for them.
```

tklauser commented on pkg/endpoint/endpoint.go:
```diff
@@ -79,6 +79,7 @@ const (
 
 	// PropertySkipBPFPolicy will mark the endpoint to skip ebpf
 	// policy regeneration.
+	// Endpoint is still placed in the endpoints map  so that ARP and ND work.
```
super nit:

```suggestion
	// Endpoint is still placed in the endpoints map so that ARP and ND work.
```

tklauser reviewed (approved):
LGTM, thanks.

FWIW, failing "Build commits" workflow should be fixed on rebase now that https://github.com/cilium/cilium/pull/35141 was merged.

julianwiedmann commented on pkg/endpoint/bpf.go:
```diff
@@ -396,6 +396,16 @@ func (e *Endpoint) regenerateBPF(regenContext *regenerationContext) (revnum uint
+				return 0, fmt.Errorf("Exposing endpoint in endpoints BPF map failed: %w", err)
+			}
+		}
+
```
Maybe update the comment, since this isn't really specific to IPv6 ND anymore ?

This code path looks completely tuned to what we currently need for the Ingress endpoint :smile:. But abstracting it is a future problem ...

jrajahalme replied:
Updated to:
```
	// Ingress endpoint needs entries in the endpoints map so that the return traffic,
	// ARP, and IPv6 ND are delivered to the host stack in all datapath configurations.
```

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -325,7 +325,9 @@ struct tunnel_value {
 } __packed;
 
-#define ENDPOINT_F_HOST		1 /* Special endpoint representing local host */
+#define ENDPOINT_F_HOST			1 /* Special endpoint representing local host */
+#define ENDPOINT_F_ATHOSTNS		2 /* Endpoint located at the host networking namespace */
+#define ENDPOINT_MASK_HOST_DELIVERY	(ENDPOINT_F_HOST | ENDPOINT_F_ATHOSTNS)
```
Nice, I like the mask approach for compatibility!

Future task will be to switch over the existing `cilium_host` endpoints to set `ENDPOINT_F_ATHOSTNS`, but not `ENDPOINT_F_HOST`. And then remove the special handling in `icmp6.h`.

julianwiedmann reviewed (commented):
Thank you Jarno, looks good! Some minor remarks below.

I'd suggest to split the change into two patches:
1. introduce `ENDPOINT_F_ATHOSTNS`, `ENDPOINT_MASK_HOST_DELIVERY` etc and motivate why it's needed.
2. switch the Ingress endpoint to set `ENDPOINT_F_ATHOSTNS`.

Besides that, could you please update the PR / patch description to reflect the latest changes? Right now it's still very much focused on the IPv6 ND problem ...

jrajahalme committed:
bpf: Add endpoint flag ENDPOINT_F_ATHOSTNS

Add a new endpoint flag ENDPOINT_F_ATHOSTNS that informs the datapath to
deliver endpoint's traffic to the host stack, but for which ARP and IPv6
ND are still served by the bpf datapath due to the endpoint's IP being
from the node's IP allocation CIDR range.

This can be used for enabling IPv6 ND for Ingress IPs in future.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Place endpoints without bpf to endpoints map

Even endpoints without bpf programs or policy need ARP and IPv6 ND to
work so that traffic can reach them. Place such endpoints in the bpf
endpoints map (aka lxcmap).

This fixes missing IPv6 ND responses for the Ingress IPs.

Fixes: #32980
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

julianwiedmann reviewed (approved):
ty Jarno, looks good! We should also ping https://github.com/cilium/cilium/issues/35277 about testing once this is merged.
